### PR TITLE
[JENKINS-31462] GitHub Enterprise Servers validation

### PR DESCRIPTION
**Pull request by recena**
[JENKINS-31462](https://issues.jenkins-ci.org/browse/JENKINS-31462)

Please, consider the comments and reviews in this PR https://github.com/jenkinsci/github-branch-source-plugin/pull/10 (original)

downstream https://github.com/jenkinsci/github-api-plugin/pull/6, downstream https://github.com/kohsuke/github-api/pull/232

#### Result

jenkins-31462

@reviewbybees, specially @jglick

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
:bee: , but the PR is blocked by github-api

**Comment by recena** [comment]
@oleg-nenashev Thanks, I'm waiting for https://github.com/kohsuke/github-api/pull/232

**Comment by recena** [comment]
@oleg-nenashev The dependency was upgraded.

File: pom.xml [hunk 1]
```diff
@@ -36,10 +36,11 @@
             <artifactId>git</artifactId>
             <version>2.3</version>
         </dependency>
+        <!-- TODO: Update when org.jenkins-ci.plugins:github-api is released -->
```
Comment (amuniz) [inline pom.xml]: It's no more a TODO

**Comment by amuniz** [comment]
:bee:

**Comment by oleg-nenashev** [comment]
Occasionally commented in the PR. :bug: for the loss of the error info in the Web UI validation output

**Comment by recena** [comment]
@oleg-nenashev 

1. I've decided to use `API URL` because GitHub API Documentation use `URL` instead of `URI`
1. I'm showing a helpful message to the use, enough to know that something is wrong

**Comment by oleg-nenashev** [comment]
I'm removing the bug according to my discussion with @recena 
<code>GitHub.connectToEnterpriseAnonymously(api.toString())</code> may throw IOException (e.g. if GitHub server is unavailable), so I still don't think the current validation is good enough to put a bee.

**Comment by oleg-nenashev** [comment]
> I'm showing a helpful message to the use, enough to know that something is wrong

The message is definitely helpful, but it does not cover all possible cases. Examples of other causes, which may lead to the IOException
* API URL is valid, but GutHub server is unavailable or shut down
* API URL is valid, but GitHub instance rejects anonymous connection attempts

**Comment by recena** [comment]
@oleg-nenashev 

>  API URL is valid, but GitHub server is unavailable or shut down

It should be considered in the library, not here.

>  API URL is valid, but GitHub instance rejects anonymous connection attempts

If you mean GitHub Enterprise in private mode, it is considered.

**Comment by oleg-nenashev** [comment]
But the current code will show "This does not look like a GitHub Enterprise API URL" in both cases, won't it? What prevents you from displaying the exception message in the validation output?

**Comment by recena** [comment]
@oleg-nenashev I hope you agree with me on both cases are exceptional. The current PR handles:

1. Empty URL
1. Invalid URL
1. A valid URL that does not correspond with a GitHub API URL

By the way, I had to modify the code after your first :bee: because @kohsuke [modified](https://github.com/kohsuke/github-api/pull/232#issuecomment-160995222) my initial proposal. I was confident with the first approach.

**Comment by recena** [comment]
@oleg-nenashev Trying to improve the PR I've found an example about why don't use `e.getMessage()` in validation messages when everything is throw in `IOException`:

```
Dec 03, 2015 7:09:14 PM org.jenkinsci.plugins.github_branch_source.Endpoint$DesciptorImpl doCheckApiUri
WARNING: Unexpected character ('<' (code 60)): expected a valid value (number, String, array, object, 'true', 'false' or 'null')
```
Imagine the warning message in the UI.

**Comment by recena** [comment]
@jglick If you agree, I'd like to merge this PR. IMO is ready to merge.

**Comment by oleg-nenashev** [comment]
I'm not convinced, but feel free to go forward if you feel strong

**Comment by recena** [comment]
@oleg-nenashev I totally disagree with show to the user technical information like `WARNING: Unexpected character ('<' (code 60)): expected a valid value (number, String, array, object, 'true', 'false' or 'null')`.

Additionally, all cases that you mentioned before should be handle in the library. I only needed a `true` or `false` but @kohsuke changed the method.

**Comment by recena** [comment]
@oleg-nenashev I agree with show to the user detailed information if he can do anything with this information. Only in these cases.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Endpoint.java [hunk 2]
```diff
@@ -89,29 +99,37 @@ public int hashCode() {
...
-            if (!enterprise.matcher(value).matches()) {
-                return FormValidation.warning(
-                        "This does not look like a GitHub Enterprise API URI. Expecting something like "
-                                + "https://[hostname or ip]/api/v3/");
+            try {
+                URL api = new URL(apiUri);
+                GitHub github = GitHub.connectToEnterpriseAnonymously(api.toString());
+                github.checkApiUrlValidity();
+                return FormValidation.ok();
+            } catch (MalformedURLException mue) {
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Endpoint.java]: You could use `catch (MalformedURLException | JsonParseException e)` and avoid the duplicated message.

**Comment by amuniz** [comment]
:bee:

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Endpoint.java (hunk 2 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Endpoint.java]: I know but we are not using this [feature](http://docs.oracle.com/javase/7/docs/technotes/guides/language/catch-multiple.html) of Java SE 7 in this plugin.

**Comment by jglick** [comment]
:bee:

**Comment by recena** [comment]
@reviewbybees done

**Pull request merged**